calyx-backend/src/verilog.rs

```diff
@@ -90,6 +94,217 @@ fn validate_control(ctrl: &ir::Control) -> CalyxResult<()> {
     }
 }
 
+/// Each external library has its own specific handler to build all information that Morty needs to do pickling. We can implement each command line argument (see https://github.com/pulp-platform/morty/blob/master/src/main.rs for available arguments) as a trait's method.
+trait LibraryHandlerTrait {
+    /// Add search path(s) for SystemVerilog includes to build a LibraryBundle
+    fn add_incs(&self) -> CalyxResult<Vec<PathBuf>>;
+    /// Directory(s) to search for SystemVerilog modules
+    fn add_library_dirs(&self) -> CalyxResult<Vec<PathBuf>>;
+    /// Define preprocesor macro(s)
+    fn add_defs(&self) -> CalyxResult<HashMap<String, Option<String>>>;
+    /// Add search path(s) for SystemVerilog includes to build a FileBundle
+    fn add_stdin_incdirs(&self) -> CalyxResult<Vec<PathBuf>>;
+    /// Add export include directories
+    fn add_export_incdirs(&self) -> CalyxResult<HashMap<String, Vec<String>>>;
+    /// Create a map from module name to file path
+    fn map_module_names_to_file_paths(
+        &self,
+    ) -> CalyxResult<HashMap<String, PathBuf>> {
+        // a hashmap from 'module name' to 'path' for all libraries.
+        let mut library_files = HashMap::new();
+        // a list of paths for all library files
+        let mut library_paths: Vec<PathBuf> = Vec::new();
+
+        // we first accumulate all library files from the 'library_dir' and 'library_file' options into
+        // a vector of paths, and then construct the library hashmap.
+        let library_dirs: Vec<PathBuf> = self.add_library_dirs()?;
+        for dir in library_dirs {
+            let entries = std::fs::read_dir(&dir).map_err(|e| {
+                Error::invalid_file(format!(
+                    "Error accessing library directory `{:?}`: {}",
+                    dir, e
+                ))
+            })?;
+
+            for entry in entries {
+                let entry = entry.map_err(|e| {
+                    Error::invalid_file(format!(
+                        "Error reading entry in directory `{:?}`: {}",
+                        dir, e
+                    ))
+                })?;
+                library_paths.push(entry.path());
+            }
+        }
+
+        for p in &library_paths {
+            // Must have the library extension (.v or .sv).
+            if morty::has_libext(p) {
+                if let Some(m) = morty::lib_module(p) {
+                    library_files.insert(m, p.to_owned());
+                }
+            }
+        }
+
+        Ok(library_files)
+    }
+}
+
+struct HardFloatHandler;
+impl LibraryHandlerTrait for HardFloatHandler {
+    fn add_incs(&self) -> CalyxResult<Vec<PathBuf>> {
```

Comment:
if you were to switch from a trait to a struct, you could include this code in a "factory" function which creates an instance of the struct